sway-core/src/semantic_analysis/ast_node/mod.rs

```diff
@@ -142,41 +142,51 @@ impl ty::TyAstNode {
                     ty::TyAstNodeContent::Declaration(ty::TyDecl::type_check(handler, ctx, decl)?)
                 }
                 AstNodeContent::Expression(expr) => {
-                    let ctx = ctx
-                        .with_type_annotation(type_engine.insert(engines, TypeInfo::Unknown, None))
-                        .with_help_text("");
+                    let mut ctx = ctx.with_help_text("");
+                    match expr.kind {
+                        ExpressionKind::ImplicitReturn(_) => {
+                            // Do not use any type annotation with implicit returns as that
+                            // will later cause type inference errors when matching implicit block
+                            // types.
+                        }
+                        _ => {
+                            ctx = ctx.with_type_annotation(type_engine.insert(
+                                engines,
+                                TypeInfo::Unknown,
+                                None,
+                            ));
+                        }
+                    }
                     let inner = ty::TyExpression::type_check(handler, ctx, expr.clone())
                         .unwrap_or_else(|err| ty::TyExpression::error(err, expr.span(), engines));
                     ty::TyAstNodeContent::Expression(inner)
                 }
-                AstNodeContent::ImplicitReturnExpression(expr) => {
-                    let ctx =
-                        ctx.with_help_text("Implicit return must match up with block's type.");
-                    let typed_expr = ty::TyExpression::type_check(handler, ctx, expr.clone())
-                        .unwrap_or_else(|err| ty::TyExpression::error(err, expr.span(), engines));
-                    ty::TyAstNodeContent::ImplicitReturnExpression(typed_expr)
-                }
                 AstNodeContent::Error(spans, err) => ty::TyAstNodeContent::Error(spans, err),
             },
             span: node.span,
         };
 
         if let ty::TyAstNode {
-            content: ty::TyAstNodeContent::Expression(ty::TyExpression { .. }),
+            content: ty::TyAstNodeContent::Expression(ty::TyExpression { expression, .. }),
             ..
-        } = node
+        } = node.clone()
```

Comment:
Here we can just borrow and avoid cloning. 
```suggestion
        } &node
```